src/FSharp.AWS.DynamoDB/StreamContext.fs

```diff
@@ -0,0 +1,419 @@
+namespace FSharp.AWS.DynamoDB
+
+open System
+open System.Threading
+open System.Threading.Tasks
+
+open Amazon.DynamoDBv2
+open Amazon.DynamoDBv2.Model
+
+type StreamOperation =
+    | Insert
+    | Modify
+    | Remove
+
+type StreamRecord<'TRecord> =
+    { TableKey: TableKey
+      Operation: StreamOperation
+      ApproximateCreationDateTime: DateTimeOffset
+      New: 'TRecord option
+      Old: 'TRecord option }
+
+type StreamPosition = { ShardId: string; SequenceNumber: string option }
+
+type ReadStreamFrom =
+    | Oldest
+    | Newest
+    | Positions of StreamPosition list
+
+module private Shard =
+
+    /// <summary>
+    /// Returns all of the open shards (with no ending sequence number) in the stream
+    /// </summary>
+    /// <param name="shards">sequence of `Shard` objects from the `StreamDescription`</param>
+    let openShards (shards: Shard seq) = shards |> Seq.filter (fun s -> String.IsNullOrEmpty(s.SequenceNumberRange.EndingSequenceNumber))
+
+    /// <summary>
+    /// Returns the ancestor shard ids of a shard in the stream
+    /// </summary>
+    /// <param name="shardId">shard id of the shard to find ancestors for</param>
+    /// <param name="shards">sequence of `Shard` objects from the `StreamDescription`</param>
+    let ancestors (shards: Shard seq) (shardId: string) =
+        let rec ancestorsInner (shardId: string) =
+            match shards |> Seq.tryFind (fun s -> s.ShardId = shardId) with
+            | None -> []
+            | Some s ->
+                match s.ParentShardId with
+                | null
+                | "" -> []
+                | parent -> parent :: ancestorsInner parent
+
+        ancestorsInner shardId |> Seq.distinct
+
+    /// <summary>
+    /// Returns true if the shard id is in the list of shards
+    /// </summary>
+    /// <param name="shards">sequence of `Shard` objects from the `StreamDescription`</param>
+    /// <param name="shardId">shard id to check for</param>
+    let containsShardId (shards: Shard seq) (shardId: string) = shards |> Seq.exists (fun s -> s.ShardId = shardId)
+
+type private ShardMsg =
+    | UpdatedShardPosition of StreamPosition
+    | EndOfShard of string
+    | ShardsUpdated of Shard seq
```

Comment:
unless there's a critical benefit to this being lazy, make ti a readonlycollection or array